pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -383,11 +383,38 @@ stdenv.mkDerivation ({
       mv tmp "$d"
     done
 
-    for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
-      for lib in "$d/"*.{dylib,so}; do
-        ln -s "$lib" "$dynamicLinksDir"
+    # This function makes the links of shared libraries this package uses,
+    # and it avoids creating duplicate links for the same file.
+    # $packageConfDir has conf files for each haskell package this package uses.
+    # There are the links of shared libraries in dynamic-library-dirs of $packageConfDir.
+    # The links are duplicated in some haskell packages.
+    makeDynamicLinks () {
+      for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
+        for lib in "$d/"*.{dylib,so} ; do
+          # When '*.{dylib,so}' does not match anything, '*.{dylib,so}' becomes '*.dylib or *.so' including '*'
+          # [ -f "$lib" ] checks if the matched path is available.
+          if [ -f "$lib" ] ; then
+            local linkPath="$dynamicLinksDir/$(basename "$lib")"
+            if [ -e "$linkPath" ]; then
+              # link already exists! but maybe that's OK
+              if [ ! -L "$linkPath" ] || [ "$(readlink -f "$linkPath")" != "$(readlink -f "$lib")" ]; then
+                echo >&2 "error: failed to create symbolic link $linkPath: file exists"
```

Comment:
I suggest explicitly not making this error look like a standard file exists error, this only seems misleading. How about "Not attempting to link from <x> to <y> because <reason>" instead